### DISCUSSION TITLE
Reverse lookup

### DISCUSSION THREAD

**Discussion by XBagon** [ozankasikci/fantasy-in-rust#23] [General]
Thought about implementing reverse lookup for chords and most importantly scales. Not really sure how to do it, if it's easier to just check all the possibilities (go through the `semitones` array) of all the current available modes, or if there's already a cool formula from music theory which allows to identify or rather **build** the term for any set of notes.

**Comment by henryksloan** [comment]
I think the most "proper" approach from a CS perspective would be a tree, where each edge represents an `Interval` added on top, and nodes represents an `Option<Chord name>`. The chord names can be generic suffixes like "minor seventh", which can be appended to the root name. I would add that the output from reverse chord/scale search should be a valid string for input, e.g. `D major seventh`.

Any reasonable approach would require a difference/subtraction operation on notes and/or pitches, so that's a good starting point. E.g. `C-A = Minor third`.

This would also mix well with the regex extensions from #22, which I now realize should include terms like "dominant seventh", and new chords like "minor sixth". After all this and that is settled, and accidentals are matured, it would be nice to support altered chords like `Cmaj7#11`.

**Reply by henryksloan** [reply]
I realized my proposal would be inflexible to inversions. Python's Mingus library seems to use interval list definitions for chord types, but does so exhaustively for all inversions of the chord. They also have separate functions for 3-note chords, etc., which could be avoided by using a tree as I recommended.

An extensive solution in a popular python library: https://github.com/bspaans/python-mingus/blob/master/mingus/core/chords.py

Stack exchange links"
https://music.stackexchange.com/q/12341
https://stackoverflow.com/q/17955486/7401492

**Reply by XBagon** [reply]
I also first thought of trees, but when we really want to support every possible scale and chord the tree would get enormous. At that point checking its properties independently from each one, would be better? I'm just not sure how the theory looks for scales. For chords this would be a lot easier and I think a tree wouldn't be necessary, I could be wrong here though. I'd imagine checking the root note, if checking its relation to the other notes, the result doesn't really depend on the last step. I guess my problem is mostly that chords are much better defined than scales are, so to find the scales we currently support a tree would be a sensible choice, but if we want to accept most (all?) combinations (which I would appreciate very much), the tree won't do much, we need The Scale Formula™️ 😄.

**Reply by henryksloan** [reply]
Fair enough, I agree that a more symbolic description may be better in the long run. However, I think if we look at the fundamental structure that lets us name scales and chords, it requires either a lookup table or a tree (though I'm open to alternatives). The methods we've mentioned so far are 1) tree and 2) properties. Though humans identify chords by looking at qualities, I'm just not sure that can be elegantly translated into an algorithm. My surface-level feeling is that a sufficiently advanced property-based method would approach a tree.

For example, imagine the parsing of the notes `C Eb Gb`. `C -> Eb` is a minor 3rd, so I imagine an algorithm would immediately identify it as a minor chord (as that is generally the principle component in identifying a chord). But then, `C -> Gb` is a diminished 5th, which is an exclusive property of diminished chords (i.e. diminished triads, o7 chords, and half-diminished chords). _That is_, unless the `Gb` were spelled as an `F#`, and were the third note in the chord `C E F# G B`, which should be called `Cmaj7#11`. As far as I can tell, in both the mental model and theoretical algorithm, we have to look at every interval w.r.t. the root in tandem - not in isolation - to identify a chord. The ordered combination of those intervals is best modeled as a tree, in my view.

I'm not totally sure how this should be done, so please let me know of any alternative algorithms you can think of. I'm sure there are solutions better than the one I'm laying out, so we should flesh out a solution together.

Also, let's keep in mind that any solution will have to account for inversions, and I think Mingus' solution of attempting to name every inversion of the given notes is appropriate. Additionally, we could "sort" the notes while keeping the given root, then test every inversion, thus allowing us to be voicing-agnostic in our naming.
